sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -1691,45 +1691,79 @@ Exceptions:
 
 |===
 
-=== Features Still in Development
-
 ==== Memory Allocation Nodes
 
-There is no provided interface for users to define a USM allocation/free
-operation belonging to the scope of the graph. It would be error prone and
-non-performant to allocate or free memory as a node executed during graph
-submission. Instead, such a memory allocation API needs to provide a way to
-return a pointer which won't be valid until the allocation is made on graph
-finalization, as allocating at finalization is the only way to benefit from
-the known graph scope for optimal memory allocation, and even optimize to
-eliminate some allocations entirely.
-
-Such a deferred allocation strategy presents challenges however, and as a result
-we recommend instead that prior to graph construction users perform core SYCL
-USM allocations to be used in the graph submission. Before to coming to this
-recommendation we considered the following explicit graph building interfaces
-for adding a memory allocation owned by the graph:
-
-1. Allocation function returning a reference to the raw pointer, i.e. `void*&`,
-   which will be instantiated on graph finalization with the location of the
-   allocated USM memory.
-
-2. Allocation function returning a handle to the allocation. Applications use
-   the handle in node command-group functions to access memory when allocated.
-
-3. Allocation function returning a pointer to a virtual allocation, only backed
-   with an actual allocation when graph is finalized or submitted.
-
-Design 1) has the drawback of forcing users to keep the user pointer variable
-alive so that the reference is valid, which is unintuitive and is likely to
-result in bugs.
-
-Design 2) introduces a handle object which has the advantages of being a less
-error prone way to provide the pointer to the deferred allocation. However, it
-requires kernel changes and introduces an overhead above the raw pointers that
-are the advantage of USM.
-
-Design 3) needs specific backend support for deferred allocation.
+Support depends on the availablity of backend support for deferred allocation:
+link:../experimental/sycl_ext_oneapi_virtual_mem.asciidoc[sycl_ext_oneapi_virtual_mem]
+
+The following interfaces enables users to define a memory allocation/free operation 
+belonging to the scope of the graph. It would be error prone and non-performant
+to allocate or free memory as a node executed during graph submission. Instead,
+such a memory allocation API needs to provide a way to return a pointer which
+won't be valid until the allocation is made on graph finalization, as allocating
+at finalization is the only way to benefit from the known graph scope for optimal
+memory allocation, and even optimize to eliminate some allocations entirely.
+
+Table {counter: tableNumber}. Member functions of the `command_graph` class (memory allocation).
+[cols="2a,a"]
+|===
+|Member function|Description
+
+|
+[source, c++]
+----
+std::pair<void*,node>
```

Comment:
I'm not sure about the concept of an allocation being a node, `sycl_ext_oneapi_virtual_mem` doesn't have allocation as being asynchronous, neither are the UR APIs - https://oneapi-src.github.io/unified-runtime/core/api.html#virtual-memory or L0 APIs - https://spec.oneapi.io/level-zero/latest/core/api.html#memory-functions

i.e. I'm not sure how we'll express node edges.

Reply:
It might be useful from a potential future optimization standpoint if say `alloc` and `free` are both nodes so that it is clear when the memory is and isn't needed, even if we are at liberty to do the actual allocation and freeing whenever we want. From an execution standpoint they can be ignored effectively.

Reply:
That's a fair point, in the future you could do more optimal memory allocations with that lifetime information.

Reply:
> That's a fair point, in the future you could do more optimal memory allocations with that lifetime information.

Yes correct, that would be the rationale behind the current draft. If we just return a `ptr` we would lose that lifetime information. Another option certainly would just tie the memory resource to a graph, which could prevent optimizations. If a node handle is returned it can be easily passed to other nodes as a dependency via its properties.

I haven't added any details on when the allocation happens, we could still do it when finalizing the graph.

> i.e. I'm not sure how we'll express node edges.

Internal representation could be similar to an empty node.

Reply:
Makes sense 👍 resolving this thread